openapps/csensors/csensors.c

```diff
@@ -0,0 +1,398 @@
+   csensors_vars.desc.componentID      = COMPONENT_CSENSORS;
+   csensors_vars.desc.discoverable     = TRUE;
+   csensors_vars.desc.callbackRx       = &csensors_receive;
+   csensors_vars.desc.callbackSendDone = &csensors_sendDone;
+   // register with the CoAP module
+   opencoap_register(&csensors_vars.desc);
+
+   opensensors = opensensors_read();
+   csensors_vars.numSensors = opensensors->numSensors;
+
+   for(i=0;i<csensors_vars.numSensors;i++) {
+      csensors_vars.csensors_resource[i].timerId = MAX_NUM_TIMERS;
+      csensors_vars.csensors_resource[i].period  = 0;
+      csensors_register(&csensors_vars.csensors_resource[i],&opensensors->opensensors_resource[i]);
+   }
+}
+
+//=========================== private =========================================
+
+void csensors_register(
+      csensors_resource_t* csensors_resource,
+      opensensors_resource_desc_t* opensensors_resource_desc
+   ) {
+   csensors_resource->desc.path0len         = sizeof(csensors_path0)-1;
+   csensors_resource->desc.path0val         = (uint8_t*)(&csensors_path0);
+   csensors_resource->desc.path1len         = opensensors_resource_desc->path1len;
+   csensors_resource->desc.path1val         = opensensors_resource_desc->path1val;
+   csensors_resource->desc.componentID      = COMPONENT_CSENSORS;
+   csensors_resource->desc.discoverable     = TRUE;
+   csensors_resource->desc.callbackRx       = &csensors_receive;
+   csensors_resource->desc.callbackSendDone = &csensors_sendDone;
+
+   // register with the CoAP module
+   opencoap_register(&csensors_resource->desc);
+}
+
+/**
+\brief Called when a CoAP message is received for this resource.
+
+\param[in] msg          The received message. CoAP header and options already
+   parsed.
+\param[in] coap_header  The CoAP header contained in the message.
+\param[in] coap_options The CoAP options contained in the message.
+
+\return Whether the response is prepared successfully.
+*/
+owerror_t csensors_receive(
+      OpenQueueEntry_t* msg,
+      coap_header_iht*  coap_header,
+      coap_option_iht*  coap_options
+   ) {
+   owerror_t outcome;
+   uint8_t   id,i;
+   uint32_t  period;
+
+
+   switch (coap_header->Code) {
+      case COAP_CODE_REQ_GET:
+         // reset packet payload
+         msg->payload                     = &(msg->packet[127]);
+         msg->length                      = 0;
+
+         if (coap_options[1].type != COAP_OPTION_NUM_URIPATH) {
+
+            // have CoAP module write links to csensors resources
+            opencoap_writeLinks(msg,COMPONENT_CSENSORS);
+
+            packetfunctions_reserveHeaderSize(msg,1);
+            msg->payload[0]     = COAP_PAYLOAD_MARKER;
+
+            // add return option
+            packetfunctions_reserveHeaderSize(msg,2);
+            msg->payload[0]     = COAP_OPTION_NUM_CONTENTFORMAT << 4 | 1;
+            msg->payload[1]     = COAP_MEDTYPE_APPLINKFORMAT;
+
+         } else {
+
+            for(id=0;id<csensors_vars.numSensors;id++) {
```

Comment:
I would call a getter for numsensors in the opensensor driver here, rather than caching the value and being outdated